RFCs/0000-pdk-template-v2.md

```diff
@@ -0,0 +1,296 @@
+
+**Template Authors**
+
+* Template authors should be able to easily add/modify templates with clear and documented standards e.g. file locations and good practices
+
+* Template authors should be able to easily generate document from their templates instead of manually keeping the code and documentation in sync.
+
+* Template authors should be able to easily test their templates work as expected, both manually and via automation, like rspec.
+
+* Template authors should be able to advertise what their templates do
+
+* Template authors should be able to use ERB for simple templates, but also put more advanced code into Ruby
+
+### Backwards compatibility
+
+The proposed changes must either be backwards compatible, or be segregated so that Template Users can progressively opt in to version 2.0
+
+## Proposed Implementation Design
+
+### Opt-in instead of opt-out
+
+Currently the templating will apply everything, and Template Users must opt-out.  The new templating system will use a composition model, that is, Template User must opt-in to things.
+
+### Easier to discover, and validate, template settings
+
+Currently Template User need to read the README to determine what settings are available or even valid. This also assumes the README is even correct.  The new templating system will have a metadata, and optionally, a validation layer so that users can discover what templates are available to opt-in to, and optionally, to validate template settings.
+
+### Easier to manage template file locations
+
+The new templating design will have descriptive and well documented file locations
+
+
+**V1 locations**
+
+| Path                      | Description |
+| ------------------------- | ----------- |
+| `/moduleroot`             | Templates get deployed on new module, convert and update |
+| `/moduleroot_init`        | Templates get only deployed when the target module does not yet exist |
+| `/object_templates/*.erb` | Template location for `pdk new ...` generation |
+| `/robocop`                | Custom ruby code to assist in generating rubocop profiles |
+
+**V2 locations**
+
+Where `name` is a unique name for the template
+
+> The `name` should use snake case (lowercase with underscore) to be cross platform compliant. The more descriptive name can be specified in the template.yml file
+
+| Path                               | Description |
+| ---------------------------------- | ----------- |
+| `/<name>/README.md`            (1) | A plain text description of the template. Though not really necessary as most of this would be contained in the `template.yml` file. It is expected that this file will be auto-generated |
+| `/<name>/template.yml`             | YAML file describing the template |
+| `/<name>/template_schema.json` (1) | JSON Schema file describing the valid settings for this template |
+| `/<name>/files/**/*`               | Templates to process for new module, convert and update |
```

Comment:
in modules, ERB and EPP templates are stored in a directory called 'templates/', while untemplated files are stored in `files/`. This should follow the same convention (although I'm not certain we need untemplated files).

Reply:
Do the templates have to follow the same naming standards? While they are related, they're very different things.